nsdb-cluster/src/main/scala/io/radicalbit/nsdb/cluster/actor/AbstractClusterListener.scala

```diff
@@ -0,0 +1,250 @@
+        .mapTo[CommitLogCoordinatorUnSubscribed]
+      _ <- (writeCoordinator ? UnSubscribePublisher(nodeName)).mapTo[PublisherUnSubscribed]
+      _ <- (writeCoordinator ? UnsubscribeMetricsDataActor(nodeName))
+        .mapTo[MetricsDataActorUnSubscribed]
+      _ <- (metadataCoordinator ? UnsubscribeMetricsDataActor(nodeName))
+        .mapTo[MetricsDataActorUnSubscribed]
+      _ <- (metadataCoordinator ? UnSubscribeCommitLogCoordinator(nodeName))
+        .mapTo[CommitLogCoordinatorUnSubscribed]
+      removeNodeMetadataResponse <- (metadataCoordinator ? RemoveNodeMetadata(nodeName))
+        .mapTo[RemoveNodeMetadataResponse]
+    } yield removeNodeMetadataResponse)
+      .retry(delay, retries)(_.isInstanceOf[NodeMetadataRemoved])
+      .map(onRemoveNodeMetadataResponse)
+  }
+
+  def receive: Receive = {
+    case MemberUp(member) if member == cluster.selfMember =>
+      log.info("Member is Up: {}", member.address)
+
+      val nodeActorsGuardian = createNodeActorsGuardian()
+
+      (nodeActorsGuardian ? GetNodeChildActors)
+        .map {
+          case NodeChildActorsGot(metadataCoordinator, writeCoordinator, readCoordinator, publisherActor) =>
+            val locationsToAdd: Seq[LocationWithCoordinates] = retrieveLocationsToAdd
+
+            val locationsGroupedBy: Map[(String, String), Seq[LocationWithCoordinates]] = locationsToAdd.groupBy {
+              case LocationWithCoordinates(database, namespace, _) => (database, namespace)
+            }
+
+            implicit val scheduler: Scheduler = context.system.scheduler
+            implicit val _log: LoggingAdapter = log
+
+            Future
+              .sequence {
+                locationsGroupedBy.map {
+                  case ((db, namespace), locations) =>
+                    metadataCoordinator ? AddLocations(db, namespace, locations.map {
+                      case LocationWithCoordinates(_, _, location) => location
+                    })
+                }
+              }
+              .map(ErrorManagementUtils.partitionResponses[LocationsAdded, AddLocationsFailed])
+              .retry(delay, retries) {
+                case (_, addLocationsFailedList) => addLocationsFailedList.isEmpty
+              }
+              .onComplete {
+                case Success((_, failures)) if failures.isEmpty =>
+                  val nodeName = createNodeName(member)
+                  mediator ! Subscribe(NODE_GUARDIANS_TOPIC, nodeActorsGuardian)
+                  mediator ! Publish(NSDB_LISTENERS_TOPIC, NodeAlive(nodeId, nodeName))
+                  NSDbClusterSnapshot(context.system).addNode(nodeId, nodeName)
+                  onSuccessBehaviour(readCoordinator, writeCoordinator, metadataCoordinator, publisherActor)
+                case e =>
+                  onFailureBehaviour(member, e)
+              }
+          case unknownResponse =>
+            log.error(s"unknown response from nodeActorsGuardian ? GetNodeChildActors $unknownResponse")
```

Comment:
should we compensate here somehow?

Reply:
Yes, good point. Agreed. I think that if a `NodeActorGuardian` returns an unknown response for a `GetChildren` request, it means that a severe issue is happening, which prevents the node from bootstrapping properly.  
Therefore, the right compensation here is shutting down the node.

Reply:
👍